src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -2146,26 +2154,37 @@ void Session::send_json_error_message()
     REALM_ASSERT(!m_unbind_message_sent);
     REALM_ASSERT(m_error_to_send);
     REALM_ASSERT(m_client_error);
+    REALM_ASSERT(!m_wait_for_error_timer);
 
     ClientProtocol& protocol = m_conn.get_client_protocol();
     OutputBuffer& out = m_conn.get_output_buffer();
     session_ident_type session_ident = get_ident();
-    std::error_code ec = m_client_error->code();
+    auto client_error = m_client_error->code();
+    auto protocol_error = static_cast<int>(client_error_to_protocol_error(client_error));
     auto message = m_client_error->what();
 
-    logger.info("Sending: ERROR \"%1\" (error_code=%2, session_ident=%3)", message, ec.value(),
+    logger.info("Sending: ERROR \"%1\" (error_code=%2, session_ident=%3)", message, protocol_error,
                 session_ident); // Throws
 
     nlohmann::json error_body_json;
     error_body_json["message"] = message;
-    protocol.make_json_error_message(out, session_ident, ec.value(), error_body_json.dump()); // Throws
-
+    protocol.make_json_error_message(out, session_ident, protocol_error, error_body_json.dump()); // Throws
     m_conn.initiate_write_message(out, this); // Throws
 
     m_error_to_send = false;
-    m_connection_to_close = true;
 
-    // Connection is waiting to be closed
+    logger.debug("Waiting for %1 milliseconds to get the ERROR message from the server", default_error_timeout);
+    m_wait_for_error_timer =
+        get_client().create_timer(std::chrono::milliseconds(default_error_timeout), [this](Status status) {
+            if (status == ErrorCodes::OperationAborted)
+                return;
+            else if (!status.is_ok())
+                throw ExceptionForStatus(status);
+
+            m_wait_for_error_timer.reset();
+            close_connection();
```

Comment:
Should we report some sort of error if the protocol error times out when trying to send to the server? Maybe through a debug message?

Reply:
good idea. I'll add a debug log